065-topic-replication-factor-change.md

```diff
@@ -0,0 +1,298 @@
+
+The goal of this proposal is to allow Strimzi users to change the topic replication factor by simply updating the KafkaTopic resource.
+Unless specified, we always refer to the Unidirectional Topic Operator implementation.
+
+## Current situation
+
+When the user change the `.spec.replicas` field of a KafkaTopic, the Topic Operator fails the reconciliation, and the resource becomes not ready. 
+An error is reported in Topic Operator logs and KafkaTopic resource status.
+
+> Replication factor change not supported, but required for partitions\
+> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,\
+> 22, 23, 24]
+
+The only [documented solution](https://strimzi.io/docs/operators/latest/deploying#proc-changing-topic-replicas-str) is to use the kafka-reassign-partitions.sh tool by spinning up a separate interactive pod.
+This is far from user friendly, and the user is in charge of doing the calculations to determine the best broker to host a new replica, or which replica needs to be dropped.
+
+When data loss is not a problem, the user can simply delete and recreate the KafkaTopic resource with the desired replication factor.
+This is not really an option for most use cases.
+
+## Motivation
+
+The topic replication factor change feature will provide a much better cloud-native experience for Strimzi users.
+They will be able to change the topic replication factor by simply updating the KafkaTopic resource, which may also happen through a GitOps pipeline.
+
+This is a non exhaustive list of use cases that will become significantly easier:
+
+- Set a lower replication factor for non-critical topics to reduce the infrastructure costs (compute, storage and network).
+- Set a lower replication factor to deal with resource shortage (trade-off between fault-tolerance and resource utilization).
+- Set a higher replication factor for critical topics to improve durability and fault tolerance.
+
+A tighter integration with Cruise Control is also one of the long term goals of the Strimzi project. 
+
+## Proposal
+
+The replication factor change feature can be implemented by integrating the Topic Operator with Cruise Control.
+The Cluster Operator is already integrated with Cruise Control, so we can factor out the existing logic and classes.
+
+Cruise Control creates a task for each operation, such as topic configuration and cluster rebalance.
+A task is first created in the Active state, then moves to InExecution, and finally to Completed or CompletedWithError.
+At most `max.active.user.tasks` tasks can be Active at the same time (5 by default), and only one task can be InExecution at any given time.
+
+During the execution, a task is divided in one or more batches, which are then executed one after the other.
+A task may affect multiple topics, which complete or fail altogether.
```

Comment:
This could do with a little elaboration, given the batching mentioned in the previous sentence. I assume you mean:

```suggestion
If any batch fails then the task becomes `CompletedWithError` potentially leaving some topics in a changed state and others unchanged.
```

Reply:
Ok.